### PR TITLE
npe fixes for `unnecessary_null_checks`

### LINKED ISSUE #2326: NPEs in unnecessary_null_checks
1. Line `80`:

```
Thread 0 main threadNoSuchMethodError: NoSuchMethodError: The getter 'staticType' was called on null.Receiver: nullTried calling: staticType
at Object.noSuchMethod(dart:core)
at getExpectedType(unnecessary_null_checks.dart:80)
at _Visitor.visitPostfixExpression(null_check_on_nullable_type_parameter.dart:81)
at PostfixExpressionImpl.accept(ast.dart:8095)
at LinterVisitor._runSubscriptions(linter_visitor.dart:755)
at LinterVisitor.visitPostfixExpression(linter_visitor.dart:539)
at PostfixExpressionImpl.accept(ast.dart:8095)
at ReturnStatementImpl.visitChildren(ast.dart:8554)
at RecursiveAstVisitor.visitReturnStatement(visitor.dart:1170)
at LinterVisitor.visitReturnStatement(linter_visitor.dart:577)
at ReturnStatementImpl.accept(ast.dart:8550)
at NodeListImpl.accept(ast.dart:7544)
at BlockImpl.visitChildren(ast.dart:1073)
at RecursiveAstVisitor.visitBlock(visitor.dart:671)
at LinterVisitor.visitBlock(linter_visitor.dart:78)
at BlockImpl.accept(ast.dart:1069)
at BlockFunctionBodyImpl.visitChildren(ast.dart:1028)
at RecursiveAstVisitor.visitBlockFunctionBody(visitor.dart:677)
at LinterVisitor.visitBlockFunctionBody(linter_visitor.dart:84)
at BlockFunctionBodyImpl.accept(ast.dart:1024)
at MethodDeclarationImpl.visitChildren(ast.dart:6976)
at RecursiveAstVisitor.visitMethodDeclaration(visitor.dart:1067)
at LinterVisitor.visitMethodDeclaration(linter_visitor.dart:474)
at MethodDeclarationImpl.accept(ast.dart:6967)
at NodeListImpl.accept(ast.dart:7544)
at ClassDeclarationImpl.visitChildren(ast.dart:1551)
at RecursiveAstVisitor.visitClassDeclaration(visitor.dart:707)
at LinterVisitor.visitClassDeclaration(linter_visitor.dart:114)
at ClassDeclarationImpl.accept(ast.dart:1521)
at NodeListImpl.accept(ast.dart:7544)
at CompilationUnitImpl.visitChildren(ast.dart:2119)
at RecursiveAstVisitor.visitCompilationUnit(visitor.dart:731)
at LinterVisitor.visitCompilationUnit(linter_visitor.dart:138)
at CompilationUnitImpl.accept(ast.dart:2112)
at LibraryAnalyzer._computeLints(library_analyzer.dart:366)
at LibraryAnalyzer.analyzeSync(library_analyzer.dart:176)
at LibraryAnalyzer.analyze(library_analyzer.dart:107)
at AnalysisDriver._computeAnalysisResult2.<fn>(driver.dart:1317)
at PerformanceLog.run(performance_logger.dart:32)
at AnalysisDriver._computeAnalysisResult2(driver.dart:1294)
at AnalysisDriver._computeAnalysisResult.<fn>(driver.dart:1247)
at runZoned(dart:async)
at NullSafetyUnderstandingFlag.enableNullSafetyTypes(null_safety_understanding_flag.dart:42)
at AnalysisDriver._computeAnalysisResult(driver.dart:1246)
at AnalysisDriver.performWork(driver.dart:1089)
at <asynchronous gap>(async)
at AnalysisDriverScheduler._run(driver.dart:2013)
at <asynchronous gap>(async)
at AnalysisDriverScheduler.start(driver.dart:1936)
at SocketServer.createAnalysisServer(socket_server.dart:82)
at StdioAnalysisServer.serveStdio(stdio_server.dart:32)
at Driver.startAnalysisServer.<fn>(driver.dart:616)
at runZoned(dart:async)
at Driver._captureExceptions(driver.dart:735)
at Driver.startAnalysisServer(driver.dart:612)
at Driver.start(driver.dart:500)
at main(main.dart:44)
```

```dart
class A {
  m3a(int? p) => p!; // OK
  m3b(int? p) {
    return p!; // OK
  }
}
```

2. Line `99`
```
Thread 0 main threadNoSuchMethodError: NoSuchMethodError: The getter 'parameters' was called on null.Receiver: nullTried calling: parameters
at Object.noSuchMethod(dart:core)
at getExpectedType(unnecessary_null_checks.dart:99)
at _Visitor.visitPostfixExpression(null_check_on_nullable_type_parameter.dart:81)
at PostfixExpressionImpl.accept(ast.dart:8095)
at LinterVisitor._runSubscriptions(linter_visitor.dart:755)
at LinterVisitor.visitPostfixExpression(linter_visitor.dart:539)
at PostfixExpressionImpl.accept(ast.dart:8095)
at BinaryExpressionImpl.visitChildren(ast.dart:960)
at RecursiveAstVisitor.visitBinaryExpression(visitor.dart:665)
at LinterVisitor.visitBinaryExpression(linter_visitor.dart:72)
at BinaryExpressionImpl.accept(ast.dart:955)
at VariableDeclarationImpl.visitChildren(ast.dart:10482)
at RecursiveAstVisitor.visitVariableDeclaration(visitor.dart:1308)
at LinterVisitor.visitVariableDeclaration(linter_visitor.dart:715)
at VariableDeclarationImpl.accept(ast.dart:10476)
at NodeListImpl.accept(ast.dart:7544)
at VariableDeclarationListImpl.visitChildren(ast.dart:10582)
at RecursiveAstVisitor.visitVariableDeclarationList(visitor.dart:1314)
at LinterVisitor.visitVariableDeclarationList(linter_visitor.dart:721)
at VariableDeclarationListImpl.accept(ast.dart:10576)
at VariableDeclarationStatementImpl.visitChildren(ast.dart:10630)
at RecursiveAstVisitor.visitVariableDeclarationStatement(visitor.dart:1320)
at LinterVisitor.visitVariableDeclarationStatement(linter_visitor.dart:727)
at VariableDeclarationStatementImpl.accept(ast.dart:10626)
at NodeListImpl.accept(ast.dart:7544)
at BlockImpl.visitChildren(ast.dart:1073)
at RecursiveAstVisitor.visitBlock(visitor.dart:671)
at LinterVisitor.visitBlock(linter_visitor.dart:78)
at BlockImpl.accept(ast.dart:1069)
at BlockFunctionBodyImpl.visitChildren(ast.dart:1028)
at RecursiveAstVisitor.visitBlockFunctionBody(visitor.dart:677)
at LinterVisitor.visitBlockFunctionBody(linter_visitor.dart:84)
at BlockFunctionBodyImpl.accept(ast.dart:1024)
at MethodDeclarationImpl.visitChildren(ast.dart:6976)
at RecursiveAstVisitor.visitMethodDeclaration(visitor.dart:1067)
at LinterVisitor.visitMethodDeclaration(linter_visitor.dart:474)
at MethodDeclarationImpl.accept(ast.dart:6967)
at NodeListImpl.accept(ast.dart:7544)
at ClassDeclarationImpl.visitChildren(ast.dart:1551)
at RecursiveAstVisitor.visitClassDeclaration(visitor.dart:707)
at LinterVisitor.visitClassDeclaration(linter_visitor.dart:114)
at ClassDeclarationImpl.accept(ast.dart:1521)
at NodeListImpl.accept(ast.dart:7544)
at CompilationUnitImpl.visitChildren(ast.dart:2119)
at RecursiveAstVisitor.visitCompilationUnit(visitor.dart:731)
at LinterVisitor.visitCompilationUnit(linter_visitor.dart:138)
at CompilationUnitImpl.accept(ast.dart:2112)
at LibraryAnalyzer._computeLints(library_analyzer.dart:366)
at LibraryAnalyzer.analyzeSync(library_analyzer.dart:176)
at LibraryAnalyzer.analyze(library_analyzer.dart:107)
at AnalysisDriver._computeAnalysisResult2.<fn>(driver.dart:1317)
at PerformanceLog.run(performance_logger.dart:32)
at AnalysisDriver._computeAnalysisResult2(driver.dart:1294)
at AnalysisDriver._computeAnalysisResult.<fn>(driver.dart:1247)
at runZoned(dart:async)
at NullSafetyUnderstandingFlag.enableNullSafetyTypes(null_safety_understanding_flag.dart:42)
at AnalysisDriver._computeAnalysisResult(driver.dart:1246)
at AnalysisDriver.performWork(driver.dart:1089)
at <asynchronous gap>(async)
at AnalysisDriverScheduler._run(driver.dart:2013)
at <asynchronous gap>(async)
at AnalysisDriverScheduler.start(driver.dart:1936)
at SocketServer.createAnalysisServer(socket_server.dart:82)
at StdioAnalysisServer.serveStdio(stdio_server.dart:32)
at Driver.startAnalysisServer.<fn>(driver.dart:616)
at runZoned(dart:async)
at Driver._captureExceptions(driver.dart:735)
at Driver.startAnalysisServer(driver.dart:612)
at Driver.start(driver.dart:500)
at main(main.dart:44)
```

Probably
```dart
hasNoOperatorPlus + p!
```

@pq @a14n

### PR DESCRIPTION
**Pull request by pq**
Fixes #2326 

/cc @scheglov @bwilkerson @a14n

### CODE REVIEW

File: test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart [hunk 1]
```diff
@@ -54,3 +54,14 @@ f5(int? p) {
...
+
+class B {
+  m3a(int? p) => p!; // OK
+  m3b(int? p) {
+    return p!; // OK
+  }
+}
+
+f6(int? p) {
+  return B() + p!; //OK
```
Comment (scheglov) [inline test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart:66]: Space after `//`?

File: test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart [hunk 2]
```diff
@@ -54,3 +54,14 @@ f5(int? p) {
   int? v1;
   v1 ??= p!; // OK
 }
+
+class B {
+  m3a(int? p) => p!; // OK
```
Comment (scheglov) [inline test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart:59]: Are these cases different from methods in `class A`?

File: test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart [hunk 3]
```diff
@@ -54,3 +54,14 @@ f5(int? p) {
...
+  m3b(int? p) {
+    return p!; // OK
+  }
+}
+
+f6(int? p) {
```
Comment (scheglov) [inline test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart:65]: Why do we put `f6` not after `f5`?
I may guess that we want to group unresolved / error cases at hte end?

File: lib/src/rules/unnecessary_null_checks.dart [hunk 4]
```diff
@@ -96,7 +96,7 @@ DartType getExpectedType(PostfixExpression node) {
   }
   // as right member of binary operator
   if (parent is BinaryExpression && parent.rightOperand == realNode) {
-    return parent.staticElement.parameters.first.type;
+    return parent.staticElement?.parameters?.first?.type;
```
Comment (scheglov) [inline lib/src/rules/unnecessary_null_checks.dart:99]: It looks like a big hammer to me, especially here, but also the change above.
Exactly one thing can be `null` here - the `staticElement`.
Using cascading `?.` hides this fact.
Maybe we should ask the element, and branch on if it is `null`.

File: test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart (hunk 3 above)
Comment (pq) [reply test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart:65]: No good reason.  It was just the order I added them. 😄 

Re-sorted.

File: test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart (hunk 2 above)
Comment (pq) [reply test/rules/experiments/nnbd/rules/unnecessary_null_checks.dart]: Nope.   Consolidated.

Thanks!

File: lib/src/rules/unnecessary_null_checks.dart (hunk 4 above)
Comment (pq) [reply lib/src/rules/unnecessary_null_checks.dart]: The one above is clumsy to re-write but this  one for sure.

Comment (pq) [reply lib/src/rules/unnecessary_null_checks.dart]: Actually.  I can fix.

**Review by scheglov — APPROVED** [review]

**Pull request merged**